File: AkarPersamaan/octave/bisection.m
function root = bisection(func, x1, x2, do_filter, tol)
% do_filter: use singularity filter or not

if nargin < 5
  tol = 1.0*eps;
end

if nargin < 4
  do_filter = false;
end

f1 = feval(func, x1);
if f1 == 0.0
  root = x1;
  return
end

f2 = feval(func, x2);
if f2 == 0.0
  root = x2;
  return
end

if f1*f2 > 0
  error('Root is not bracketed in (x1,x2)')
end

MaxIter = ceil( log(abs(x2-x1)/tol)/log(2.0) );
fprintf('MaxIter = %d\n', MaxIter)

for i = 1:MaxIter
  x3 = 0.5*(x1 + x2);
  f3 = feval(func, x3);
  %
  if( (do_filter == true) && (abs(f3) > abs(f1)) && (abs(f3) > abs(f2) ) )
    root = NaN;
    return
  end
  %
  if f3 == 0.0
    root = x3;
    return
  end
  %
  if f2*f3 < 0.0
    x1 = x3;
    f1 = f3;
  else
    x2 = x3;
    f2 = f3;
  end
end

root = 0.5*(x1 + x2);

File: AkarPersamaan/python3/bisection.py
from math import log10, ceil

def bisection(f, x1, x2, TOL=1.0e-9, verbose=False, NiterMax=None ):

    if verbose:
        print("")
        print("Searching root with bisection method:")
        print("Interval = (%18.10f,%18.10f)" % (x1,x2))
        print("TOL = %e" % TOL)
        print("")

    f1 = f(x1)
    if abs(f1) <= TOL:
        return x1, 0.0

    f2 = f(x2)
    if abs(f2) <= TOL:
        return x2, 0.0

    if f1*f2 > 0.0:
        raise RuntimeError("Root is not bracketed")

    # No NiterMax is provided
    # We calculate the default value here.
    if NiterMax == None:
        NiterMax = int(ceil( log10(abs(x2-x1)/TOL) )/ log10(2.0) ) + 10
        # extra 10 iterations

    # For the purpose of calculating relative error
    x3 = 0.0
    x3_old = 0.0

    if verbose:
        print(13*" "+"Iter      Estimated          f(x)")
        print(13*" "+"----      ---------          ----")
        print("")


    for i in range(1,NiterMax+1):

        x3_old = x3
        x3 = 0.5*(x1 + x2)
        f3 = f(x3)

        if verbose:
            print("bisection: %5d %18.10f %15.5e" % (i, x3, abs(f3)))

        if abs(f3) <= TOL:
            if verbose:
                print("")
                print("bisection is converged in %d iterations" % i)
            # return the result
            return x3, abs(x3 - x3_old)


        if f2*f3 < 0.0:
            # sign of f2 and f3 is different
            # root is in [x2,x3]
            # change the interval bound of x1 to x3
            x1 = x3
            f1 = f3
        else:
            # sign of f1 and f3 is different
            # root is in [x1,x3]
            # change the interval bound of x2 to x3
            x2 = x3
            f2 = f3


    # No root is found after NiterMax iterations
    if verbose:
        print("No root is found")
    return None, None

